File: frontend/src/features/mail/providers/MailProvider.ts
import type { MailAttachment, MailFolder, MailIdentity, MailMessage, MailSearchQuery, MailThread } from '../types';

export type ProviderType = 'ews' | 'gmail' | 'imap' | 'jmap';

/** Capabilities advertised by a mail source. UI code must consult this
 * contract instead of inferring support from the provider type. */
export interface MailProviderCapabilities {
  readonly snooze: boolean;
  readonly scheduledSend: {
    readonly supported: boolean;
    readonly maxDelaySeconds?: number;
  };
}

export interface MailItemRef {
  item_id: string;
  change_key: string;
  conversation_id?: string;
}

/** A file attached by the user in the composer, ready to be sent. */
export interface ComposerAttachment {
  name: string;
  contentType: string;
  size: number;
  /** Base64-encoded file content (no data URL prefix). */
  data: string;
  /** True for images embedded inline in the HTML body (referenced via CID). */
  isInline?: boolean;
  /** Content-ID for inline images (without angle brackets), matches src="cid:…" in the HTML. */
  contentId?: string;
}

export interface SendMailParams {
  to: string[];
  cc?: string[];
  bcc?: string[];
  subject: string;
  bodyHtml: string;
  replyToItemId?: string | null;
  replyToChangeKey?: string | null;
  isForward?: boolean;
  attachments?: ComposerAttachment[];
  fromIdentityId?: string;
  /** RFC 5322 In-Reply-To value (Message-ID of the parent, with angle brackets). */
  inReplyTo?: string;
  /** RFC 5322 References value (space-separated chain of Message-IDs). */
  references?: string;
  /** Server-side delivery date (ISO 8601). Omit for immediate delivery. */
  sendAt?: string;
}

export interface SaveDraftParams {
  to: string[];
  cc?: string[];
  bcc?: string[];
  subject: string;
  bodyHtml: string;
}

export interface Contact {
  email: string;
  name?: string;
  source?: 'google-contact' | 'google-other-contact' | 'ews-contact' | 'ews-directory' | 'mail' | 'calendar';
}

export interface ContactBackfillBatch {
  observations: Array<{
    email: string;
    displayName?: string;
    kind: 'received' | 'sent';
    occurredAt: number;
    eventId: string;
  }>;
  itemCount: number;
  oldestAt?: number;
}

/**
 * Abstraction over any mail backend (EWS, Gmail, …).
 * Add a new implementation to support additional providers without touching the UI.
 */
export interface MailProvider {
  readonly providerType: ProviderType;
  readonly accountId: string;
  readonly capabilities: MailProviderCapabilities;
  /** Resolve account/server-dependent capabilities (notably JMAP limits). */
  getCapabilities?(): Promise<MailProviderCapabilities>;

  listThreads(folder: string, maxCount?: number, offset?: number): Promise<MailThread[]>;
  /** Total number of conversations in a folder, when the protocol exposes it. */
  getThreadCount?(folder: string): Promise<number>;
  /** Load a preview lazily when listThreads intentionally omits it. */
  getThreadSnippet?(conversationId: string): Promise<string>;
  /** Search threads using a structured query. Results are not cached. */
  searchThreads(query: MailSearchQuery, maxCount?: number): Promise<MailThread[]>;
  getThread(conversationId: string, includeTrash?: boolean, isDraft?: boolean, includeDrafts?: boolean): Promise<MailMessage[]>;
  /** Lazily load a single message body. Providers may omit this when getThread already returns bodies. */
  getMessageContent?(messageId: string, conversationId?: string): Promise<Pick<MailMessage, 'body_html' | 'body_text' | 'ics_mime' | 'attachments' | 'has_attachments'>>;
  /** Return the original RFC 5322/MIME source when supported by the provider. */
  getRawMessageSource?(itemId: string): Promise<string>;
  listFolders(): Promise<MailFolder[]>;
  sendMail(params: SendMailParams): Promise<void>;
  markRead(items: MailItemRef[]): Promise<void>;
  markUnread(items: MailItemRef[]): Promise<void>;
  moveToTrash(itemId: string): Promise<void>;
  permanentlyDelete(itemId: string): Promise<void>;
  /** Batch trash — prefer over N individual moveToTrash calls to avoid rate-limits. */
  bulkMoveToTrash(conversationIds: string[]): Promise<void>;
  /** Batch permanent delete — prefer over N individual permanentlyDelete calls. */
  bulkPermanentlyDelete(conversationIds: string[]): Promise<void>;
  /** Batch folder move — prefer over N individual moveToFolder calls. */
  bulkMoveToFolder(conversationIds: string[], folderId: string): Promise<void>;
  openAttachment(attachment: MailAttachment): Promise<void>;
  /** Return the attachment content as a standard base64 string (for in-app preview / download). */
  getAttachmentData(attachment: MailAttachment): Promise<string>;
  saveDraft(params: SaveDraftParams): Promise<string>;
  findOrCreateSnoozedFolder?(): Promise<string>;
  moveToFolder(itemId: string, folderId: string): Promise<void>;
  snooze?(itemId: string, until?: string): Promise<string>;
  getInboxUnread(): Promise<number>;
  listIdentities?(): Promise<MailIdentity[]>;
  searchContacts?(query: string, maxCount?: number): Promise<Contact[]>;
  /** Returns the contact photo as a base64 string (no data-URL prefix), or null if unavailable. */
  getContactPhoto?(email: string): Promise<string | null>;
  /** Efficient provider-native metadata scan used by the persistent contact index. */
  backfillContacts?(folder: string, offset: number, maxCount: number): Promise<ContactBackfillBatch>;
}

File: frontend/src-tauri/src/mail_provider.rs
use serde::{Deserialize, Serialize};

// ── Canonical shared types ─────────────────────────────────────────────────────

#[derive(Serialize, Debug, Clone)]
pub struct MailThread {
    pub conversation_id: String,
    pub topic: String,
    pub snippet: String,
    pub last_delivery_time: String,
    pub message_count: u32,
    pub unread_count: u32,
    pub from_name: Option<String>,
    pub has_attachments: bool,
}

#[derive(Serialize, Debug, Clone)]
pub struct MailMessage {
    pub item_id: String,
    pub change_key: String,
    pub subject: String,
    pub from_name: Option<String>,
    pub from_email: Option<String>,
    pub to_recipients: Vec<MailRecipient>,
    pub cc_recipients: Vec<MailRecipient>,
    pub body_html: String,
    pub date_time_received: String,
    pub is_read: bool,
    pub has_attachments: bool,
    pub attachments: Vec<MailAttachment>,
    /// ICS text extracted from a calendar MIME part (EWS meeting requests).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub ics_mime: Option<String>,
    /// True when the message is a draft (EWS only).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub is_draft: Option<bool>,
    /// Plain-text body for quote-marker detection.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub body_text: Option<String>,
    /// RFC 5322 Message-ID (JMAP only, used for threading).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub message_id: Option<String>,
    /// RFC 5322 References header (JMAP only, used for threading).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub references: Option<String>,
}

#[derive(Serialize, Debug, Clone)]
pub struct MailRecipient {
    pub name: Option<String>,
    pub email: String,
}

#[derive(Serialize, Debug, Clone)]
pub struct MailFolder {
    pub folder_id: String,
    pub display_name: String,
    pub total_count: u32,
    pub unread_count: u32,
}

#[derive(Serialize, Debug, Clone)]
pub struct MailAttachment {
    pub attachment_id: String,
    pub name: String,
    pub content_type: String,
    pub size: u64,
    pub is_inline: bool,
}

/// Minimal item reference used by mark-read / mark-unread / move commands.
/// `folder` is never sent from TypeScript — it is set by the Rust IMAP adapter
/// before calling the provider, since IMAP requires a folder context.
#[derive(Deserialize, Debug)]
pub struct MailItemRef {
    pub item_id: String,
    pub change_key: String,
    pub conversation_id: Option<String>,
    #[serde(skip)]
    pub folder: Option<String>,
}

/// A file attachment supplied by the frontend composer (base64-encoded content).
#[derive(Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct ComposerAttachment {
    pub name: String,
    pub content_type: String,
    pub data: String,
    pub is_inline: Option<bool>,
    pub content_id: Option<String>,
}

/// Structured search query forwarded from the TypeScript provider.
#[derive(Deserialize, Debug)]
pub struct MailSearchQuery {
    pub from: Option<String>,
    pub to: Option<String>,
    pub cc: Option<String>,
    pub bcc: Option<String>,
    pub subject: Option<String>,
    /// Free-text search in body/subject.
    pub text: Option<String>,
    /// Well-known folder key or arbitrary provider folder ID.
    pub folder: Option<String>,
    /// 'today', 'yesterday', or 'YYYY-MM-DD'.
    pub date: Option<String>,
}

#[derive(Serialize, Debug, Clone)]
pub struct MailIdentity {
    pub id: String,
    pub name: String,
    pub email: String,
    pub may_delete: bool,
}

pub struct SendMailParams {
    pub to: Vec<String>,
    pub cc: Vec<String>,
    pub bcc: Vec<String>,
    pub subject: String,
    pub body_html: String,
    /// Item to reply to or forward (EWS only).
    pub reply_to_item_id: Option<String>,
    pub reply_to_change_key: Option<String>,
    pub attachments: Option<Vec<ComposerAttachment>>,
    pub is_forward: Option<bool>,
    /// Sending identity (JMAP only).
    pub identity_id: Option<String>,
    /// RFC 5322 In-Reply-To value (JMAP only).
    pub in_reply_to: Option<String>,
    /// RFC 5322 References chain (JMAP only).
    pub references: Option<String>,
}

pub struct SaveDraftParams {
    pub to: Vec<String>,
    pub cc: Vec<String>,
    pub bcc: Vec<String>,
    pub subject: String,
    pub body_html: String,
}

// ── Trait ─────────────────────────────────────────────────────────────────────

/// Contract that every mail backend must satisfy.
/// Add a new provider by creating a struct and implementing this trait.
/// Methods with default bodies return `Err("not_supported")` — override only what
/// the backend actually supports.
pub trait MailProvider {
    // ── Required ──────────────────────────────────────────────────────────────

    async fn list_folders(&self) -> Result<Vec<MailFolder>, String>;
    async fn get_inbox_unread(&self) -> Result<u32, String>;
    async fn list_threads(&self, folder: &str, max_count: Option<u32>) -> Result<Vec<MailThread>, String>;
    async fn get_thread(
        &self,
        conversation_id: &str,
        include_trash: Option<bool>,
        is_draft: Option<bool>,
        include_drafts: Option<bool>,
    ) -> Result<Vec<MailMessage>, String>;
    async fn mark_read(&self, items: &[MailItemRef]) -> Result<(), String>;
    async fn mark_unread(&self, items: &[MailItemRef]) -> Result<(), String>;
    async fn move_to_trash(&self, item_id: &str) -> Result<(), String>;
    async fn permanently_delete(&self, item_id: &str) -> Result<(), String>;
    async fn bulk_move_to_trash(&self, item_ids: Vec<String>) -> Result<(), String>;
    async fn bulk_permanently_delete(&self, item_ids: Vec<String>) -> Result<(), String>;
    async fn send_mail(&self, params: SendMailParams) -> Result<(), String>;
    /// `message_id` and `folder` are only used by IMAP (which needs both to locate
    /// the body section). EWS and JMAP pass `None` for these.
    async fn get_attachment_data(
        &self,
        attachment_id: &str,
        message_id: Option<&str>,
        folder: Option<&str>,
    ) -> Result<String, String>;

    // ── Optional (default: not_supported) ─────────────────────────────────────

    async fn search_threads(&self, _query: &MailSearchQuery, _max_count: Option<u32>) -> Result<Vec<MailThread>, String> {
        Err("not_supported".to_string())
    }
    async fn save_draft(&self, _params: SaveDraftParams) -> Result<String, String> {
        Err("not_supported".to_string())
    }
    async fn open_attachment(&self, _attachment_id: &str, _filename: &str) -> Result<(), String> {
        Err("not_supported".to_string())
    }
    async fn move_to_folder(&self, _item_id: &str, _folder_id: &str) -> Result<(), String> {
        Err("not_supported".to_string())
    }
    async fn bulk_move_to_folder(&self, _item_ids: Vec<String>, _folder_id: &str) -> Result<(), String> {
        Err("not_supported".to_string())
    }
    async fn find_or_create_snoozed_folder(&self) -> Result<String, String> {
        Err("not_supported".to_string())
    }
    async fn snooze(&self, _item_id: &str) -> Result<String, String> {
        Err("not_supported".to_string())
    }
    async fn list_identities(&self) -> Result<Vec<MailIdentity>, String> {
        Err("not_supported".to_string())
    }
}
